### DISCUSSION TITLE
[Modules Support] Password Support

### DISCUSSION THREAD

**Discussion by Spiegie** [linkdd/tricorder#26] [Ideas]
In Configuration, Passwords are often a Problem.

If you apply a Module via commanline on your local machine, I don't think, we have a problem with passwords. 
The encrypted password can be stored in a Vault or something, and you could decrypt it by passing a password or a key. 
The Problem emerges, If you want to hold up a State by the Module. You could do that, by running the state every hour or so. That needs to be possible automatically.

I would love it, if a Module could apply itself on the host. But I don't have much hope, especially because I love the agentless nature of this project. 

That's why I just want to ask if you (the reader) got any idea, how we could implement passwords.

**Comment by linkdd** [comment]
The inventory can be dynamic. If it is an executable, then `tricorder` will execute it and expect the inventory to be returned as a JSON on stdout.

Therefore, if you want to have passwords in your inventory variables, you should use a dynamic inventory which will then read the credentials from somewhere else (environment variables? hashicorp vault? etc...).

By not supporting natively the encryption of the inventory (like ansible does with `ansible-vault`), we keep things simple, and give more power to the user (which can then use anything they want to solve that issue).

Imagine this in pseudo-python:

```python
import vault
import json
import os

vault_token = os.getenv("VAULT_TOKEN")
api_credentials = vault.read("/api/credentials", token=vault_token)
api_user = api_credentials["user"]
api_password = api_credentials["password"]

inventory = {
  "hosts": [
    {
      "id": "foo",
      "address": "10.0.1.3:22",
      "user": "root",
      "tags": ["foobar"],
      "vars": {"user": api_user, "password": api_password},
    },
  ],
}

print(json.dumps(inventory))
```

Finally, in your CI/CD pipeline (be it with Gitlab CI, or Github Actions, or anything else), you can configure a secret variable "VAULT_TOKEN" that is exposed to your pipeline via environment variables.

Or you can skip Hashicorp Vault, and use CI/CD secrets (which I would not recommend if you are paranoid about security).

**Comment by Spiegie** [comment]
I think your point is very Valid. Keeping it simple and giving the Power to the user should be more in my focus.
